frontend/packages/console-shared/src/modals/modal-resource-launcher.tsx

```diff
@@ -0,0 +1,87 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import * as _ from 'lodash';
+import { Provider } from 'react-redux';
+import { Router } from 'react-router-dom';
+
+import store from '@console/internal/redux';
+import { Firehose, AccessDenied, history, MsgBox, Box } from '@console/internal/components/utils';
+import { ModalErrorContent } from '@console/internal/components/modals/error-modal';
+
+const NotFound = ({ message }) => (
+  <Box className="text-center">
+    <MsgBox title="Not Found" detail="Requested resource not found." />
+    {_.isString(message) && (
+      <div className="alert alert-danger text-left">
+        <span className="pficon pficon-error-circle-o" />
+        {message}
+      </div>
+    )}
+  </Box>
+);
+
+const ModalComponentWrapper = (props: ComponentWrapperProps) => {
+  const { Component, closeModal, resourceToProps, loadError, loaded, ...restProps } = props;
+
+  if (loadError) {
+    const status = _.get(loadError, 'response.status');
+    let errorContent;
+    if (status === 404) {
+      errorContent = <NotFound message={loadError.message} />;
+    }
+    if (status === 403) {
+      errorContent = <AccessDenied message={loadError.message} />;
+    }
+
+    if (!loaded) {
+      return <ModalErrorContent cancel={closeModal} error={errorContent} />;
+    }
+  }
+
+  return (
+    <Component
+      onClose={closeModal}
+      onCancel={closeModal}
+      onHide={closeModal}
+      {...props}
+      {...resourceToProps({ ...restProps })}
+    />
+  );
+};
+
+type ComponentWrapperProps = {
+  loadError?: any;
+  loaded?: boolean;
+  Component: React.ComponentType<any>;
+  closeModal: (e?: React.SyntheticEvent) => any;
+  resourceToProps: (res: any) => any;
+};
+
+export const modalResourceLauncher = (ModalComponent, resources, resourceToProps) => (props) => {
+  const modalContainer = document.getElementById('modal-container');
```

Comment:
It looks like a lot of this code was copied directly from factory/modal.tsx. Can we find a way to implement this without duplicate code?

Reply:
right, will try

Reply:
done, @spadgett can you please have a look?